utils/s3/client.cc

```diff
@@ -98,6 +98,191 @@ future<> client::delete_object(sstring object_name) {
+    auto req = http::request::make("PUT", _client->_host, _object_name);
+    req._headers["Content-Length"] = format("{}", bufs.size());
+    req.query_parameters["partNumber"] = format("{}", part_number);
+    req.query_parameters["uploadId"] = _upload_id;
+    req.write_body("bin", bufs.size(), [this, part_number, bufs = std::move(bufs)] (output_stream<char>&& out_) mutable -> future<> {
+        auto out = std::move(out_);
+        s3l.trace("upload {} part data (upload id {})", part_number, _upload_id);
+        for (auto&& buf : bufs.buffers()) {
+            co_await out.write(buf.get(), buf.size());
+        }
+        co_await out.flush();
+        co_await out.close();
+    });
+
+    auto units = co_await get_units(_flush_sem, 1);
+    (void)_http.make_request(std::move(req), [this, part_number] (http::reply rep, input_stream<char>&& in_) mutable -> future<> {
+        auto etag = rep.get_header("ETag");
+        s3l.trace("uploaded {} part data -> etag = {} (upload id {})", part_number, etag, _upload_id);
+        _part_etags[part_number] = std::move(etag);
+        return make_ready_future<>();
+    }).handle_exception([this, part_number] (auto ex) {
+        s3l.warn("couldn't upload part {}: {} (upload id {})", part_number, ex, _upload_id);
+    }).finally([units = std::move(units)] {});
+}
+
+future<> client::upload_sink::abort_upload() {
+    s3l.trace("DELETE upload {}", _upload_id);
+    auto req = http::request::make("DELETE", _client->_host, _object_name);
+    req.query_parameters["uploadId"] = std::move(_upload_id);
+    co_await _http.make_request(std::move(req), ignore_reply, http::reply::status_type::no_content);
+}
+
+future<> client::upload_sink::finalize_upload() {
+    co_await do_flush();
+
+    s3l.trace("wait for {} parts to complete (upload id {})", _part_etags.size(), _upload_id);
+    co_await _flush_sem.wait(flush_concurrency);
+
+    sstring parts_xml = dump_multipart_upload_parts(_part_etags);
+    if (parts_xml.empty()) {
+        co_await abort_upload();
+        co_await coroutine::return_exception(std::runtime_error("couldn't upload parts"));
+    }
+
+    s3l.trace("POST upload completion {} parts (upload id {})", _part_etags.size(), _upload_id);
+    auto req = http::request::make("POST", _client->_host, _object_name);
+    req.query_parameters["uploadId"] = std::move(_upload_id);
+    req.write_body("xml", std::move(parts_xml));
+    s3l.trace("-> [{}]", req.content);
+    co_await _http.make_request(std::move(req), ignore_reply);
+}
+
+future<> client::upload_sink::close() {
+    if (upload_started()) {
+        s3l.warn("closing incomplete multipart upload -> aborting");
+        co_await abort_upload();
+    } else {
+        s3l.trace("closing multipart upload");
+    }
+}
+
+data_sink client::make_upload_sink(sstring object_name) {
+    return data_sink(std::make_unique<upload_sink>(shared_from_this(), std::move(object_name)));
+}
+
 future<> client::close() {
```

Comment:
10,000 parts * 5MB = 50GB. Too small for comfort.

If we multiply minimum_part_size by a factor k each time, we can make it arbitrarily large.

Looks like the sum is first_part*(1-k^n)/(1-k).

My math isn't up to solving it, but experimenting shows k=1.0005 is more than enough. The last part ends up 750GB though.

Reply:
1.00045 gets us to 1TB with with "just" 445MB last part.

Reply:
This is very important, specially because we don't split large partitions yet. I have seen >200G large partition in a single sstable. I think with today's implementation, the upload will just fail as S3 will reject another part once the 10k limit is reached.

@avikivity I think you were referring to UploadPart instead, as UploadPartCopy requires using existing objects as data source.

Reply:
I was referring to UploadPartCopy. The first level uses 10,000 UploadPart * 5MB = 50GB to create several 50GB sub-objects. The second level uses UploadPartCopy to splice multiple 50GB sub-objects into a single 182TB sstable containing a single partition, and winning the large partition trophy for 2023.

Reply:
(but, we needn't do it just now)

Reply:
Ah I was focused on your first proposal of gradually increasing part (chunk) size. It's not good (throughput wise) that we have to do a second pass on the uploaded data, but I think you prefer it to avoid stressing node's memory when the input data is very very large.

Reply:
It's S3 that has to do this second pass. It's not nice but it allows keeping memory usage at 5MB.

Reply:
Is there a way to pre-estimate the final object size somehow? If the object is going to be below 50Gb, then "default" uploading would look like

```
CreateUpload(temp)
UploadPart * N
CompleteMultipartUpload(temp)
CreateUpload(obj)
UploadPartCopy(from temp)
CompleteMultipartUpload(obj)
```

while it could be 2x times shorter

Reply:
If we assume that it's under 50GB we still need to send some value in the ContentLength header, and if the real  amount of data turns out to be less than that we're in trouble :(

Reply:
If we know memtable or compaction will produce output which is smaller than 50G, then s3 client can task the fast path and it seems to me it can avoid UploadPartCopy altogether. As for compaction, it can definitely provide a deterministic response that it won't generate output smaller than 50G. Perhaps the only edge case is if user disabled compression and the output is larger than input, but we can detect it using compression ratio statistics in the SSTable and looking at the schema (if compression has been disabled).

As for ContentLength header, it doesn't really matter because we still buffer the output in 5MB chunks and flush it either after the buffer is full or operation, e.g. compaction, has exhausted the input data.